contracts/liquidator/Liquidator.sol

```diff
@@ -0,0 +1,206 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity 0.8.15;
+pragma abicoder v2;
+
+import "./vendor/@uniswap/v3-core/contracts/interfaces/callback/IUniswapV3FlashCallback.sol";
+import "./vendor/@uniswap/v3-core/contracts/libraries/LowGasSafeMath.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/base/PeripheryPayments.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/base/PeripheryImmutableState.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/libraries/PoolAddress.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/libraries/CallbackValidation.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/libraries/TransferHelper.sol";
+import "./vendor/@uniswap/v3-periphery/contracts/interfaces/ISwapRouter.sol";
+
+import "../CometInterface.sol";
+import "../ERC20.sol";
+
+contract Liquidator is IUniswapV3FlashCallback, PeripheryImmutableState, PeripheryPayments {
+    struct FlashParams {
+        address[] accounts;
+        address pairToken;
+        uint24 poolFee;
+        bool reversedPair;
+    }
+
+    struct FlashCallbackData {
+        uint256 amount;
+        address payer;
+        PoolAddress.PoolKey poolKey;
+        address[] assets;
+        uint256[] baseAmounts;
+        bool reversedPair;
+    }
+
+    using LowGasSafeMath for uint256;
+    using LowGasSafeMath for int256;
+
+    ISwapRouter public immutable swapRouter;
+    CometInterface public immutable comet;
+
+    uint24 public constant defaultPoolFee = 500;
+    mapping(address => uint24) public poolFees;
+
+    constructor(
+        ISwapRouter _swapRouter,
+        CometInterface _comet,
+        address _factory,
+        address _WETH9,
+        address[] memory _assets,
+        uint24[] memory _poolFees
+    ) PeripheryImmutableState(_factory, _WETH9) {
+        require(_assets.length == _poolFees.length, "Wrong input");
+
+        swapRouter = _swapRouter;
+        comet = _comet;
+
+        // Set the desirable pool fees for assets
+        for (uint i = 0; i < _assets.length; i++) {
+            address asset = _assets[i];
+            uint24 poolFee = _poolFees[i];
+            poolFees[asset] = poolFee;
+        }
+    }
+
+    function getPoolFee(address asset) internal view returns(uint24) {
+        uint24 poolFee = poolFees[asset];
+        return poolFee == 0 ? defaultPoolFee : poolFee;
+    }
+
+    function uniswapV3FlashCallback(
+        uint256 fee0,
+        uint256 fee1,
+        bytes calldata data
+    ) external override {
+        FlashCallbackData memory decoded = abi.decode(data, (FlashCallbackData));
+        CallbackValidation.verifyCallback(factory, decoded.poolKey);
+
+        address[] memory assets = decoded.assets;
+
+        uint256 totalAmountOut = 0;
+        for (uint i = 0; i < assets.length; i++) {
+            address asset = assets[i];
+            uint256 baseAmount = decoded.baseAmounts[i];
+
+            if (baseAmount == 0) continue;
+
+            // XXX approve everything all at once?
```

Comment:
Why not?

Reply:
no reason not to

Reply:
Do you mean approve `uint256.MAX` instead of `baseAmount`?

Reply:
Oh is that what you meant? I guess its a bit more fragile if we rely on all the tokens having the approve max behavior

Reply:
Wouldn't setting the approve amount to `type(uint256).max` pass for all ERC20 tokens? It could also save some extra gas.